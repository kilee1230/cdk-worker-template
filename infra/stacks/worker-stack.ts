import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as sns from "aws-cdk-lib/aws-sns";
import * as sqs from "aws-cdk-lib/aws-sqs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { NodejsFunction, OutputFormat } from "aws-cdk-lib/aws-lambda-nodejs";
import * as lambdaEventSources from "aws-cdk-lib/aws-lambda-event-sources";
import * as snsSubscriptions from "aws-cdk-lib/aws-sns-subscriptions";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export class WorkerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create SNS Topic
    const topic = new sns.Topic(this, "WorkerTopic", {
      displayName: "Worker Topic",
      topicName: "worker-topic",
    });

    // Create Dead Letter Queue
    const deadLetterQueue = new sqs.Queue(this, "WorkerDLQ", {
      queueName: "worker-dlq",
      retentionPeriod: cdk.Duration.days(14),
    });

    // Create Main SQS Queue with DLQ
    const queue = new sqs.Queue(this, "WorkerQueue", {
      queueName: "worker-queue",
      visibilityTimeout: cdk.Duration.seconds(300),
      receiveMessageWaitTime: cdk.Duration.seconds(20),
      deadLetterQueue: {
        queue: deadLetterQueue,
        maxReceiveCount: 3, // Move to DLQ after 3 failed attempts
      },
    });

    // Subscribe SQS Queue to SNS Topic
    topic.addSubscription(new snsSubscriptions.SqsSubscription(queue));

    // Create Lambda Function with esbuild bundling
    const workerFunction = new NodejsFunction(this, "WorkerFunction", {
      runtime: lambda.Runtime.NODEJS_20_X,
      handler: "handler",
      entry: join(__dirname, "../../src/lambda/worker/index.ts"),
      timeout: cdk.Duration.seconds(30),
      memorySize: 256,
      bundling: {
        minify: true,
        sourceMap: true,
        target: "es2022",
        format: OutputFormat.ESM,
        banner:
          '{"js":"import { createRequire } from \'module\';const require = createRequire(import.meta.url);"}',
        esbuildArgs: {
          "--tree-shaking": "true",
        },
      },
      environment: {
        QUEUE_URL: queue.queueUrl,
        TOPIC_ARN: topic.topicArn,
        DLQ_URL: deadLetterQueue.queueUrl,
        ENVIRONMENT: "production",
        LOG_LEVEL: "info",
      },
    });

    // Add SQS as event source for Lambda
    workerFunction.addEventSource(
      new lambdaEventSources.SqsEventSource(queue, {
        batchSize: 10,
        maxBatchingWindow: cdk.Duration.seconds(5),
        reportBatchItemFailures: true,
      })
    );

    // Grant permissions
    queue.grantConsumeMessages(workerFunction);
    topic.grantPublish(workerFunction);
    deadLetterQueue.grantSendMessages(workerFunction);

    // Output the resource ARNs/URLs
    new cdk.CfnOutput(this, "TopicArn", {
      value: topic.topicArn,
      description: "SNS Topic ARN",
    });

    new cdk.CfnOutput(this, "QueueUrl", {
      value: queue.queueUrl,
      description: "SQS Queue URL",
    });

    new cdk.CfnOutput(this, "DLQUrl", {
      value: deadLetterQueue.queueUrl,
      description: "Dead Letter Queue URL",
    });

    new cdk.CfnOutput(this, "FunctionName", {
      value: workerFunction.functionName,
      description: "Lambda Function Name",
    });
  }
}
