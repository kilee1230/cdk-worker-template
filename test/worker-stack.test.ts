import { describe, it, expect } from "vitest";
import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { WorkerStack } from "../infra/stacks/worker-stack.js";

describe("WorkerStack", () => {
  describe("Snapshot Tests", () => {
    it("should match the CloudFormation template snapshot", () => {
      const app = new cdk.App();
      const stack = new WorkerStack(app, "TestStack");
      const template = Template.fromStack(stack);

      // Snapshot the entire CloudFormation template
      expect(template.toJSON()).toMatchSnapshot();
    });

    it("should match the resource count snapshot", () => {
      const app = new cdk.App();
      const stack = new WorkerStack(app, "TestStack");
      const template = Template.fromStack(stack);

      const resources = template.toJSON().Resources;
      const resourceTypes = Object.entries(resources).reduce(
        (acc, [, resource]) => {
          const type = (resource as any).Type;
          acc[type] = (acc[type] || 0) + 1;
          return acc;
        },
        {} as Record<string, number>
      );

      // Snapshot resource types and counts
      expect(resourceTypes).toMatchSnapshot();
    });

    it("should match the outputs snapshot", () => {
      const app = new cdk.App();
      const stack = new WorkerStack(app, "TestStack");
      const template = Template.fromStack(stack);

      // Snapshot all stack outputs
      expect(template.toJSON().Outputs).toMatchSnapshot();
    });
  });

  it("should create SNS topic with correct properties", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that SNS topic is created
    template.resourceCountIs("AWS::SNS::Topic", 1);

    template.hasResourceProperties("AWS::SNS::Topic", {
      DisplayName: "Worker Topic",
      TopicName: "worker-topic",
    });
  });

  it("should create SQS queue with DLQ configuration", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that 2 queues are created (main queue + DLQ)
    template.resourceCountIs("AWS::SQS::Queue", 2);

    // Check main queue properties
    template.hasResourceProperties("AWS::SQS::Queue", {
      QueueName: "worker-queue",
      VisibilityTimeout: 300,
      ReceiveMessageWaitTimeSeconds: 20,
    });

    // Check DLQ properties
    template.hasResourceProperties("AWS::SQS::Queue", {
      QueueName: "worker-dlq",
      MessageRetentionPeriod: 1209600, // 14 days in seconds
    });
  });

  it("should create Lambda function with correct configuration", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that Lambda function is created
    template.resourceCountIs("AWS::Lambda::Function", 1);

    template.hasResourceProperties("AWS::Lambda::Function", {
      Runtime: "nodejs20.x",
      Handler: "index.handler",
      Timeout: 30,
      MemorySize: 256,
    });
  });

  it("should configure Lambda with environment variables", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check Lambda environment variables
    template.hasResourceProperties("AWS::Lambda::Function", {
      Environment: {
        Variables: {
          ENVIRONMENT: "production",
          LOG_LEVEL: "info",
        },
      },
    });
  });

  it("should create SQS subscription to SNS topic", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that SNS subscription is created
    template.resourceCountIs("AWS::SNS::Subscription", 1);

    template.hasResourceProperties("AWS::SNS::Subscription", {
      Protocol: "sqs",
    });
  });

  it("should create Lambda event source mapping for SQS", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that event source mapping is created
    template.resourceCountIs("AWS::Lambda::EventSourceMapping", 1);

    template.hasResourceProperties("AWS::Lambda::EventSourceMapping", {
      BatchSize: 10,
      MaximumBatchingWindowInSeconds: 5,
      FunctionResponseTypes: ["ReportBatchItemFailures"],
    });
  });

  it("should create IAM role for Lambda with proper permissions", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that IAM role is created
    template.resourceCountIs("AWS::IAM::Role", 1);

    // Verify the role has Lambda service principal
    template.hasResourceProperties("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          },
        ],
      },
    });
  });

  it("should create CloudFormation outputs", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Check that outputs are defined
    template.hasOutput("TopicArn", {});
    template.hasOutput("QueueUrl", {});
    template.hasOutput("DLQUrl", {});
    template.hasOutput("FunctionName", {});
  });

  it("should have appropriate resource count", () => {
    const app = new cdk.App();
    const stack = new WorkerStack(app, "TestStack");
    const template = Template.fromStack(stack);

    // Verify total resource counts
    const resources = template.toJSON().Resources;
    const resourceCount = Object.keys(resources).length;

    // Should have multiple resources (SNS, SQS, Lambda, IAM, etc.)
    expect(resourceCount).toBeGreaterThan(5);
  });
});
