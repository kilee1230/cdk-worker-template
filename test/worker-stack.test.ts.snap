// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkerStack > Snapshot Tests > should match the CloudFormation template snapshot 1`] = `
{
  "Outputs": {
    "DLQUrl": {
      "Description": "Dead Letter Queue URL",
      "Value": {
        "Ref": "WorkerDLQ878BA572",
      },
    },
    "FunctionName": {
      "Description": "Lambda Function Name",
      "Value": {
        "Ref": "WorkerFunctionACE6A4B0",
      },
    },
    "QueueUrl": {
      "Description": "SQS Queue URL",
      "Value": {
        "Ref": "WorkerQueueE3D427EB",
      },
    },
    "TopicArn": {
      "Description": "SNS Topic ARN",
      "Value": {
        "Ref": "WorkerTopic7BD0C79A",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "WorkerDLQ878BA572": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "worker-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "WorkerFunctionACE6A4B0": {
      "DependsOn": [
        "WorkerFunctionServiceRoleDefaultPolicy07D8276A",
        "WorkerFunctionServiceRole38D8F738",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "70cd151727c65e630a809803ca9062bdcce31c041b5873c0340bf9ef096dea36.zip",
        },
        "Environment": {
          "Variables": {
            "DLQ_URL": {
              "Ref": "WorkerDLQ878BA572",
            },
            "ENVIRONMENT": "production",
            "LOG_LEVEL": "info",
            "QUEUE_URL": {
              "Ref": "WorkerQueueE3D427EB",
            },
            "TOPIC_ARN": {
              "Ref": "WorkerTopic7BD0C79A",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "WorkerFunctionServiceRole38D8F738",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WorkerFunctionServiceRole38D8F738": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkerFunctionServiceRoleDefaultPolicy07D8276A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WorkerQueueE3D427EB",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "WorkerTopic7BD0C79A",
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WorkerDLQ878BA572",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WorkerFunctionServiceRoleDefaultPolicy07D8276A",
        "Roles": [
          {
            "Ref": "WorkerFunctionServiceRole38D8F738",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkerFunctionSqsEventSourceTestStackWorkerQueue9B1F4B3E70766AEE": {
      "Properties": {
        "BatchSize": 10,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "WorkerQueueE3D427EB",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "WorkerFunctionACE6A4B0",
        },
        "FunctionResponseTypes": [
          "ReportBatchItemFailures",
        ],
        "MaximumBatchingWindowInSeconds": 5,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "WorkerQueueE3D427EB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "worker-queue",
        "ReceiveMessageWaitTimeSeconds": 20,
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "WorkerDLQ878BA572",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "WorkerQueuePolicy0AF57CF8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "WorkerTopic7BD0C79A",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "WorkerQueueE3D427EB",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "WorkerQueueE3D427EB",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "WorkerQueueTestStackWorkerTopic1FC7471E615F6138": {
      "DependsOn": [
        "WorkerQueuePolicy0AF57CF8",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "WorkerQueueE3D427EB",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": {
          "Ref": "WorkerTopic7BD0C79A",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "WorkerTopic7BD0C79A": {
      "Properties": {
        "DisplayName": "Worker Topic",
        "TopicName": "worker-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`WorkerStack > Snapshot Tests > should match the outputs snapshot 1`] = `
{
  "DLQUrl": {
    "Description": "Dead Letter Queue URL",
    "Value": {
      "Ref": "WorkerDLQ878BA572",
    },
  },
  "FunctionName": {
    "Description": "Lambda Function Name",
    "Value": {
      "Ref": "WorkerFunctionACE6A4B0",
    },
  },
  "QueueUrl": {
    "Description": "SQS Queue URL",
    "Value": {
      "Ref": "WorkerQueueE3D427EB",
    },
  },
  "TopicArn": {
    "Description": "SNS Topic ARN",
    "Value": {
      "Ref": "WorkerTopic7BD0C79A",
    },
  },
}
`;

exports[`WorkerStack > Snapshot Tests > should match the resource count snapshot 1`] = `
{
  "AWS::IAM::Policy": 1,
  "AWS::IAM::Role": 1,
  "AWS::Lambda::EventSourceMapping": 1,
  "AWS::Lambda::Function": 1,
  "AWS::SNS::Subscription": 1,
  "AWS::SNS::Topic": 1,
  "AWS::SQS::Queue": 2,
  "AWS::SQS::QueuePolicy": 1,
}
`;
